presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkAdaptiveQueryExecution.java

```diff
@@ -0,0 +1,351 @@
+                sparkTaskDescriptorJsonCodec,
+                queryStatusInfoJsonCodec,
+                queryDataJsonCodec,
+                rddFactory,
+                transactionManager,
+                pagesSerde,
+                executionExceptionFactory,
+                queryTimeout,
+                queryCompletionDeadline,
+                metadataStorage,
+                queryStatusInfoOutputLocation,
+                queryDataOutputLocation,
+                tempStorage,
+                nodeMemoryConfig,
+                waitTimeMetrics,
+                errorClassifier,
+                planFragmenter,
+                metadata,
+                partitioningProviderManager,
+                historyBasedPlanStatisticsTracker);
+
+        this.executedFragments = new HashSet<>();
+        this.iterativePlanFragmenter = createIterativePlanFragmenter();
+        this.fragmentEventQueue = new LinkedBlockingQueue<>();
+    }
+
+    private IterativePlanFragmenter createIterativePlanFragmenter() {
+        // Function<PlanFragmentId, SubPlan> getSubPlanForId = null;
+
+        // TODO Rebecca to check: I have not found any place that this is being called with true but let me know.
+        boolean forceSingleNode = false;
+        Function<PlanFragmentId, Boolean> isFragmentFinished = this.executedFragments::contains;
+
+        // TODO The static plan fragmenter is created by injection (and is also passed in the constructor here but remains unused).
+        //  We should do the same here (the static and dynamic fragmenters need to belong to the same superclass) to also avoid
+        //  using directly constructors for various parameters of the constructor.
+        return new IterativePlanFragmenter(this.planAndMore.getPlan(), isFragmentFinished, this.metadata, new PlanChecker(new FeaturesConfig(), forceSingleNode), new SqlParser(),
+                new PlanNodeIdAllocator(), new PrestoSparkNodePartitioningManager(this.partitioningProviderManager), new QueryManagerConfig(), this.session, this.warningCollector,
+                forceSingleNode);
+    }
+
+    @Override
+    protected List<Tuple2<MutablePartitionId, PrestoSparkSerializedPage>> doExecute()
+            throws SparkException, TimeoutException
+    {
+        queryStateTimer.beginRunning();
+
+        IterativePlanFragmenter.PlanAndFragments planAndFragments = iterativePlanFragmenter.createReadySubPlans(this.planAndMore.getPlan().getRoot());
+
+        // TODO Are the parameter values okay?
+        ExecutionContextExecutorService execService = ExecutionContext.fromExecutorService(ThreadUtils.newDaemonCachedThreadPool("AdaptiveExecution", 16, 60));
+
+        while (planAndFragments.hasRemainingPlan()) {
+            List<SubPlan> readyFragments = planAndFragments.getReadyFragments();
+
+            for (SubPlan fragment : readyFragments) {
+                // TODO Is it right to create the tableWriteInfo for every fragment or just once is sufficient?
```

Comment:
just once should be good.  The table you're writing too shouldn't change with reoptimization.  Actually creating the writer target does some metadata calls (https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/execution/scheduler/TableWriteInfo.java#L74), so probably best to only do once.